posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1002 @@
+
+## Step 10: Build Operators into the Tasks
+
+We now need to import our operator, import the `os` package, and tell Airflow which directory it needs to look at for templating. To do so:
+
+1. In `change_data_capture.py`, add the following to the imports at the top of the file:
+
+    ```python
+    import os
+    from plugins.gcp_custom.operators.gcp_cloudsql_export_operator import (
+        CloudSqlCsvExportOperator
+    )
+
+    ROOT_DIR = os.getenv('AIRFLOW_HOME')
+    TEMPLATES_DIR = f'{ROOT_DIR}/include/template'
+    gcp_conn_id = 'cloud_sql_api_connection'
+    db_instance_id = 'your_db_instance_id'
+    gcp_project_id = 'your_db_project_id'
+    db_name = 'your_db_name'
+    gcs_bucket = 'your_gcs_bucket'
+    gcs_root = f'{gcs_bucket}/cloudsql_export'
+    offload_export = False # True for serverless, more expensive and longer running, but offloads completely
+    ```
+
+    >*Note: All of these strings will be added later as a JSON-serialized airflow variable, which can be configured through the Airflow UI.*
+
+ 2. Add the `template_searchpath` parameter to your `DAG` definition like so:
+
+    ```python
+    with DAG('change_data_capture_cloud_sql',
+              max_active_runs=1, # Ensure one at a time
+              schedule_interval=None, # Start with None for manual trigger, change to '@hourly'
+              catchup=False, # If True, it will start historically to start_date
+              template_searchpath=TEMPLATES_DIR, # Add this**
+              start_date=datetime(2020, 12, 29, 0, 0, 0),
+              default_args=default_args) as dag:
+    ```
+
+    With that, we can simply pass the `.sql` file path to our operators instead of the entire query.  
+
+3. Update the `get_schema` task to show the following:
+
+    ```python
+    gcs_schema_keypath = (
+        f"{gcs_root}/"
+        "schemas/"
+        "{{execution_date.year}}/"
+        "{{execution_date.month}}/"
+        "{{execution_date.day}}/"
+        f"{db_name}/"
+        f"{db_name}_schema_"
+        "{{ts_nodash}}.csv"
+    )
+
+    get_schema = CloudSqlCsvExportOperator(
+        task_id = 'get_schema',
+        gcp_conn_id = gcp_conn_id,
+        database_instance = db_instance_id,
+        project_id = gcp_project_id,
+        database_name = db_name,
+        gcs_bucket = gcs_bucket,
+        gcs_keypath = gcs_schema_keypath,
+        offload = offload_export,
+        params = {
+            "tables": tables
+        },
+        export_sql = 'cloud_sql_cdc/export_csv/get_schema.sql',
+        pool = 'cloudsql_operations',
+        trigger_rule = 'one_success'
+    )
+    ```
+
+    > *Notice here how we also used `execution_date` to construct the GCS keypath. You'll see another macro there as well `{{ts_nodash}}`, which is `execution_date.isoformat()` with special characters and spaces removed, useful for time-based object keys.*
```

Comment:
"which is the equivalent of"

Reply:
Good catch. This one is always confusing too. I feel like it's "which is the equivalent to" here because it's used as something that has value

Reply:
When equivalent is used as a noun, you use "of". When it's an adjective, you use "to".

For example:
"This is the equivalent of that"
vs.
"This is equivalent to that"

Reply:
thank you